test/profiling/alt-impl/cid1.js

```diff
@@ -0,0 +1,260 @@
+          Buffer.from(codecVarints[this.codec]),
+          this.multihash
+        ])
+      default:
+        throw new Error('unsupported version')
+    }
+  }
+
+  /**
+   * Get the prefix of the CID.
+   *
+   * @returns {Buffer}
+   * @readonly
+   */
+  get prefix () {
+    return Buffer.concat([
+      Buffer.from(`0${this.version}`, 'hex'),
+      codecVarints[this.codec],
+      multihash.prefix(this.multihash)
+    ])
+  }
+
+  /**
+   * Convert to a CID of version `0`.
+   *
+   * @returns {CID1}
+   */
+  toV0 () {
+    if (this.codec !== 'dag-pb') {
+      throw new Error('Cannot convert a non dag-pb CID to CIDv0')
+    }
+
+    return new CID1(0, this.codec, this.multihash)
+  }
+
+  /**
+   * Convert to a CID of version `1`.
+   *
+   * @returns {CID1}
+   */
+  toV1 () {
+    return new CID1(1, this.codec, this.multihash)
+  }
+
+  /**
+   * Encode the CID into a string.
+   *
+   * @param {string} [base='base58btc'] - Base encoding to use.
+   * @returns {string}
+   */
+  toBaseEncodedString (base) {
+    base = base || 'base58btc'
+
+    switch (this.version) {
+      case 0: {
+        if (base !== 'base58btc') {
+          throw new Error('not supported with CIDv0, to support different bases, please migrate the instance do CIDv1, you can do that through cid.toV1()')
+        }
+        return mh.toB58String(this.multihash)
+      }
+      case 1:
+        return multibase.encode(base, this.buffer).toString()
+      default:
+        throw new Error('Unsupported version')
+    }
+  }
+
+  /**
+   * Serialize to a plain object.
+   *
+   * @returns {SerializedCID}
+   */
+  toJSON () {
+    return {
+      codec: this.codec,
+      version: this.version,
+      hash: this.multihash
+    }
+  }
+
+  /**
+   * Compare equality with another CID.
+   *
+   * @param {CID1} other
+   * @returns {bool}
+   */
+  equals (other) {
+    return this.codec === other.codec &&
+      this.version === other.version &&
+      this.multihash.equals(other.multihash)
+  }
+
+  /**
+   * Test if the given input is a CID.
+   *
+   * @param {any} other
+   * @returns {bool}
+   */
+  static isCID (other) {
+    try {
+      CID1.validateCID(other)
+    } catch (err) {
+      return false
+    }
+
+    return true
+  }
+
+  /**
+   * Test if the given input is a valid CID object.
+   * Throws if it is not.
+   *
+   * @param {any} other
+   * @returns {void}
+   */
+  static validateCID (other) {
+    if (other == null) {
+      throw new Error('null values are not valid CIDs')
+    }
+
+    if (!(other.version === 0 || other.version === 1)) {
+      throw new Error('Invalid version, must be a number equal to 1 or 0')
+    }
+
+    if (typeof other.codec !== 'string') {
+      throw new Error('codec must be string')
+    }
+
+    if (!Buffer.isBuffer(other.multihash)) {
+      throw new Error('multihash must be a Buffer')
+    }
+
+    mh.validate(other.multihash)
+  }
+}
+
+CID1.codecs = codecs
+
+module.exports = CID1
```

Comment:
Can we call it CID_A and CID_B given that we already use 0 and 1 for version numbers? Pretty cool to include profiling tests!

Reply:
Thinking more about this. Instead of duplicating the code. This could just be installing the CID version on npm.

Reply:
Doing it this way will require creating a new npm project folder (with package.json) for testing/profiling. This only works for the "reference" impl. We may still need a separate/duplicated CID for comparison. (See the comment below.)